core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,194 @@
+from micropython import const
+
+from trezor.crypto import hashlib
+from trezor.messages import CardanoAddressType
+from trezor.messages.CardanoAddressParametersType import CardanoAddressParametersType
+
+from apps.common import cbor
+
+from ..address import derive_address_bytes, get_address_bytes_unsafe
+from ..certificates import cborize_certificate
+from .lazy_collection import DICT, LIST, LazyCollection
+
+if False:
+    from trezor.messages.CardanoSignTx import CardanoSignTx
+    from trezor.messages.CardanoTxInput import CardanoTxInput
+    from trezor.messages.CardanoTxOutput import CardanoTxOutput
+    from trezor.messages.CardanoAssetGroup import CardanoAssetGroup
+    from trezor.messages.CardanoToken import CardanoToken
+    from trezor.messages.CardanoTxCertificate import CardanoTxCertificate
+    from trezor.messages.CardanoTxWithdrawal import CardanoTxWithdrawal
+    from .. import seed
+    from typing import Any, Optional
+
+INPUT = const(0)
+OUTPUT = const(1)
+ASSET_GROUP = const(2)
+TOKEN = const(3)
+CERTIFICATE = const(4)
+WITHDRAWAL = const(5)
+
+
+class TxBuilder:
+    def __init__(self, keychain: seed.Keychain, msg: CardanoSignTx):
+        tx_body_map_item_count = 3 + sum(
+            1
+            for entry in [
+                msg.ttl,
+                msg.certificates_count,
+                msg.withdrawals_count,
+                msg.has_auxiliary_data,
+                msg.validity_interval_start,
+            ]
+            if bool(entry)
+        )
+        self.state = [LazyCollection(DICT, tx_body_map_item_count)]
+        self.serializer = cbor.encode_streamed(self.state[0])
+        self.hashfn = hashlib.blake2b(outlen=32)
+        self._update_tx_hash()
+        self.keychain = keychain
+        self.msg = msg
+        self.expected_item_kind: Optional[int] = None
+
+    def _update_tx_hash(self) -> None:
+        while (chunk := next(self.serializer)) is not None:
+            self.hashfn.update(chunk)
+
+    def get_tx_hash(self) -> bytes:
+        assert self.state[0].is_filled(), "The hash calculation is not finished yet"
+        return self.hashfn.digest()
+
+    def _start_collection_at_key(
+        self,
+        key: Any,
+        collection: LazyCollection,
+        next_kind: int,
+        current_kind: Optional[int] = None,
+    ) -> None:
+        self._add_item((key, collection), current_kind)
+        self.state.append(collection)
+        self.expected_item_kind = next_kind
+
+    def _add_item(self, item: Any, item_kind: Optional[int] = None) -> None:
```

Comment:
these underscored methods seems to me as a whole "class within a class" - wouldn't it make sense perhaps to extract them into some HashedCborBuilder "sister" class which would expose these generic (_add_item/_start_collection_at_key/...) methods and return the hash of whatever hash function is passed to its constructor (to not couple it too much with blake2b256 which can actually differ even within Cardano)

Reply:
For me, those underscored methods are the "private" implementation of the "public" interface methods.
If I extract them, I'll leave some interface class, which just delegates everything to the implementation class.
We can parameterize the hash fn on the `TxBuilder` constructor as well.

Reply:
I see the underscored methods as a different level of abstraction from the non-underscored methods - that's why I would consider splitting it and make the "outward facing" TxBuilder class smaller and hence easier to follow. I see also potential for reusing the CborBuilder in the future, e.g. if we introduce message signing functionality: https://cips.cardano.org/cips/cip8/ - but that's of course not something I would consciously prepare for now as we may as well never need it, still - I find it quite natural to split the general cbor building abstraction from the tx builing one

Parametrizing the hash function in the TxBuilder doesn't have much sense as the hash function for Cardano transactions can't really change